### DISCUSSION TITLE
Many-to-many Relation is not Detected

### DISCUSSION THREAD

**Discussion by AurelliaChristie** [supabase/supabase#14000] [Questions] [answered]
Hi there, 
I'm currently working on a project where I have these two tables: `project` & `profile` that are related. The relation is many-to-many, so I created a new table called `projectmembers` that have these two foreign keys: `projectid` & `userid` which are connected to the respective table. 

In my NextJS app, I wanted to fetch all members' detail belonging to a specific project. Based on this [docs](https://supabase.com/docs/guides/api/joins-and-nesting), I understand that we can directly skip the intermediary table (`projectmembers`) and go straight to the `profile` table, using this query code:

```
const { data, error } = await supabase
        .from('project')
        .select(
            `*, members:profile(*)`
        )
        .eq('handle', handle)
        .single();
```

But I got an error as follows when trying to use that query:

```
{
    code: 'PGRST200',
    details: "Searched for a foreign key relationship between 'project' and 'profile' in the schema 'public', but no matches were found.",
    hint: null,
    message: "Could not find a relationship between 'project' and 'profile' in the schema cache"
}
```
Here I also attach the screenshot of the public schema of the `projectmembers` table that I fetched using the SQL Query Editor.

![image](https://user-images.githubusercontent.com/69672839/234921191-5e23a5dd-b49b-4e1b-a6a3-7f04c635a60b.png)

I don't understand what I did wrong. From what I understand from the docs and other references, Supabase is supposed to be able to detect the many-to-many relationship in this kind of situation. Can anyone help me with this problem? Thank you in advance!

**Comment by steve-chavez** [comment, marked as answer]
Hello,

Sorry, the docs need updating(will follow up with a PR). For many-to-many  relationships to be detected, the primary key must be declared like this:

```sql
create table
  members (
    "id" serial,
    "user_id" int references users,
    "team_id" int references teams,
    primary key(id, user_id, team_id)
  );
```

**Reply by steve-chavez** [reply]
Just fixed this on https://github.com/supabase/supabase/pull/14006.

**Reply by AurelliaChristie** [reply]
Thank you! it works now

**Reply by gusvd** [reply]
I've been trying to create and query a many-to-many relationship and I keep getting the same error as the OP.

I've create junction table using the following:

```sql
create table
  related_symbols (
    source_symbol_id uuid references symbols,
    related_symbol_id uuid references symbols,
    primary key (source_symbol_id, related_symbol_id)
  );
```

I'm querying using:
```js
let { data, error } = await supabase
  .from("symbols")
  .select("name, symbols(name)")
  .then(({ data, error }) => {
    console.log("data", data, "\nerror:", error);
  });
```

And keep getting the following error:

```
data null 
error: {
  code: 'PGRST200',
  details: "Searched for a foreign key relationship between 'symbols' and 'symbols' in the schema 'public', but no matches were found.",
  hint: null,
  message: "Could not find a relationship between 'symbols' and 'symbols' in the schema cache"
}
```

Any hints?

**Reply by MorningLit** [reply]
@gusvd I know it's been a few months already but I faced the same issue as you, but when I use the solution that @steve-chavez provided above, eg:
```
create table
  members (
    "id" serial,
    "user_id" int references users,
    "team_id" int references teams,
    primary key(id, user_id, team_id)
  );
```
It works now, not sure why an id is needed for the joins to work, and not sure if the docs need updating again

**Reply by filippo-orru** [reply]
In my case, I forgot to add a row-level security policy (RLS) for the join table, so I was getting an empty array as a response when doing a many-to-many join. Maybe this helps someone in the future :)
